### PR TITLE
Map QM paths to host file paths

### PR DESCRIPTION
**Pull request by tomjn**
Tells qm how to adjust paths in stack traces to use the host machines file paths

See https://github.com/humanmade/altis-dev-tools/issues/16

### CODE REVIEW

File: inc/namespace.php [hunk 1]
```diff
@@ -75,4 +75,32 @@ function bootstrap() {
...
+}
+
+/**
+ * Enables Query Monitor to map paths to their
+ * original values on the host.
+ *
+ * @param array $map an associative array of folders as keys, and their mappings as values
+ * @return array the adjusted mapping of folders
+ */
+function set_file_path_map( $map ) : array {
```
Comment (pdewouters) [inline inc/namespace.php]: `array $map`?

Comment (tomjn) [reply inc/namespace.php]: agreed

File: inc/namespace.php [hunk 2]
```diff
@@ -75,4 +75,32 @@ function bootstrap() {
...
+function set_file_path_map( array $map ) : array {
+	if ( ! file_exists( '/etc/chassis-constants' ) ) {
+		return $map;
+	}
+	$json_string = file_get_contents( '/etc/chassis-constants' );
+	if ( empty( $json_string ) ) {
+		return $map;
+	}
+	$data = json_decode( $json_string, true );
+	if ( empty( $data['synced_folders'] ) ) {
```
Comment (rmccue) [inline inc/namespace.php]: I'd prefer an explicit `empty( $data )` here, as it's a different code path.
```suggestion
	if ( empty( $data ) || empty( $data['synced_folders'] ) ) {
```

File: inc/namespace.php [hunk 3]
```diff
@@ -75,4 +75,32 @@ function bootstrap() {
...
+
+	add_filter( 'qm/output/file_path_map', __NAMESPACE__ . '\\set_file_path_map', 1 );
+}
+
+/**
+ * Enables Query Monitor to map paths to their
+ * original values on the host.
+ *
+ * @param array $map an associative array of folders as keys, and their mappings as values
+ * @return array the adjusted mapping of folders
```
Comment (rmccue) [inline inc/namespace.php]: Comments should be updated per https://github.com/humanmade/altis-local-chassis/pull/39#pullrequestreview-300798375

**Comment by tomjn** [comment]
This should be fine, but, for whatever reason `composer` doesn't pick up the `local-chassis` subcommand with this branch on my local instance, likely due to the namespace change, but I'm not seeing what the culprit is 🤔

**Comment by tomjn** [comment]
`composer local-server` is working locally now, false alarm :/

File: inc/namespace.php [hunk 4]
```diff
@@ -75,4 +75,31 @@ function bootstrap() {
...
+}
+
+/**
+ * Enables Query Monitor to map paths to their original values on the host.
+ *
+ * @param array $map Map of guest path => host path
+ * @return array Adjusted mapping of folders
+ */
+function set_file_path_map( array $map ) : array {
+	if ( ! file_exists( '/etc/chassis-constants' ) ) {
```
Comment (roborourke) [inline inc/namespace.php]: Is this meant to be in the local chassis repo? This file won't exist on local server

Comment (roborourke) [reply inc/namespace.php]: Odd. I figured out an alternative method that works at least. Cheers for getting this off the ground. It’s very cool

**Pull request merged**